File: 79-Word-Search/solution.cpp
#include <vector>
#include <iostream>
#include <string>

using std::vector;
using std::string;

class Solution {
    public:
        bool exist(vector<vector<char>>& board, string word) {
            if(board.size() == 0) return false;

            for ( int i = 0; i < board.size(); ++i)
            {
                for( int j = 0; j < board[0].size(); ++j) 
                {
                    if (exist(board, word, 0, i, j)) return true;
                }
            }
            return false;
        }

    private:
        bool exist(vector<vector<char>>& board, string& word, int w, int i, int j) {
            if ( w >= word.size()) return true;
            if ( i >= board.size() || j >= board[0].size() ) return false;
            bool ret = false;
            char tmp;
            if(board[i][j] == word[w] && board[i][j] != '\0')
            {
                tmp = board[i][j]; 
                board[i][j] = '\0';
                ret = exist(board, word, w+1, i, j-1) || exist(board, word, w+1, i + 1, j) || exist(board, word, w+1, i, j+1) || exist(board, word, w+1, i - 1, j);
                board[i][j] = tmp;
            }

            return ret;
        }
};

int main() 
{
    return 0;
}

File: 79-Word-Search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(board, word, 0, i, j):
                    return True
        return False

    def dfs(self, board, word, w, i, j):
        if w == len(word):
            return True

        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]):
            return False

        if board[i][j] == word[w]:
            board[i][j] = '-'

            for next_i, next_j in [[i + 1, j], [i - 1, j], [i, j - 1],
                                   [i, j + 1]]:
                if self.dfs(board, word, w + 1, next_i, next_j):
                    return True

            board[i][j] = word[w]

        return False
